lib/elixir/lib/string.ex

```diff
@@ -1121,9 +1121,28 @@ defmodule String do
       iex> String.replace("a,b,c", ",", "[]", insert_replaced: [1, 1])
       "a[,,]b[,,]c"
 
+  When an empty string is provided as a `pattern`, the function will treat it as
+  an implicit empty string between each character and the string will be
+  interspersed. If an empty string is provided as `replacement` the `subject`
+  will be returned:
+
+      iex> String.replace("ELIXIR", "", ".")
```

Comment:
I'm wondering if it is not consistent with how `split/2` sees empty strings inside:
```elixir
String.split("ELIXIR", "")
#=> ["E", "L", "I", "X", "I", "R", ""]
```
Should not `String.replace("ELIXIR", "", ".")` return `"E.L.I.X.I.R."` then? 🤔

Reply:
@lexmag that however does not mirror how replace_suffix and replace_prefix works today. In Rust we would get an empty string before and after, so I would rather say it is a bug on how split works.

Reply:
@josevalim great.

Reply:
We can't change how split works because it is deeply grained into how split with regex works.

The other point to consider is that changing split may not be the way to go depending on how we want to write the properties. For example, this holds today:

```elixir
check all string <- string(:unicode) do
  assert length(String.split(string, "")) == String.length + 1
end
```

But if we rewrite split to have an empty string before and after, then it no longer holds. The main point here still holds though, replace_suffix, replace_prefix and replace are inconsistent.

Reply:
> But if we rewrite split to have an empty string before and after, then it no longer holds.

If todays property holds, it does not mean that it is correct though; with fixed splice we would have `+ 2` instead of `+ 1`, thus we still have a property.
Maybe we should change `Regex.split/3` as well, considering that `Regex.replace(~r//, "qwe", ".")` sees an empty string before and returns `".q.w.e."`?

Reply:
@lexmag We can generalize the property above:

```elixir
check all string <- string(:unicode), count <- integer() do
  duplicate = String.duplicate(string, count)
  assert String.split(duplicate, string)) == count + 1
end
```

The only way the property above holds is if the previous property also holds. We can generally say that the number of results in split is the number of occurrences of the pattern + 1. Adding an empty string before and after makes no sense, I am afraid.

Reply:
Btw, great catch regarding Regex.replace. That puts me clearly in the camp of merging this PR. Split and replace will have the inconsistency we have been debating but this inconsistency exists today and I would prefer to make all replacements consistent between them and all splits consistent between them.